packages/api/src/external/commonwell-v2/patient/patient.ts

```diff
@@ -622,6 +598,80 @@ async function updatePatient({
   return updatedId;
 }
 
+/**
+ * Runs the patient linking flow with retries.
+ *
+ * As we upgrade links, the search fans out to find more potential links.
+ */
+async function runPatientLinkingWithRetries({
+  commonWell,
+  patient,
+  commonwellPatientId,
+  context,
+  getOrgIdExcludeList,
+}: {
+  commonWell: CommonWellAPI;
+  patient: Patient;
+  commonwellPatientId: string;
+  context: string;
+  getOrgIdExcludeList: () => Promise<string[]>;
+}): Promise<{
+  validLinks: NetworkLink[];
+  invalidLinks: NetworkLink[];
+}> {
+  let validLinks: NetworkLink[] = [];
+  let invalidLinks: NetworkLink[] = [];
+  let attempt = 0;
+
+  while (attempt < MAX_ATTEMPTS_PATIENT_LINKING) {
+    attempt++;
+    const { log } = out(
+      `retryPatientLinkingFlow - attempt ${attempt}/${MAX_ATTEMPTS_PATIENT_LINKING} - patientId ${patient.id}`
+    );
+
+    const [existingLinks, probableLinks] = await Promise.all([
+      getExistingLinks({
+        commonWell,
+        commonwellPatientId,
+      }),
+      getProbableLinks({
+        commonWell,
+        commonwellPatientId,
+      }),
+    ]);
+
+    const existingLinksCount = existingLinks?.Patients?.length ?? 0;
+    const probableLinksCount = probableLinks?.Patients?.length ?? 0;
+    log(
+      `Found ${existingLinksCount} existing links, and ${probableLinksCount} probable links on attempt ${attempt}`
+    );
+
+    const result = await tryToImproveLinks({
+      commonWell,
+      patient,
+      commonwellPatientId,
+      existingLinks,
+      probableLinks,
+      context,
+      getOrgIdExcludeList,
+    });
+
+    validLinks = result.validLinks;
+    invalidLinks = result.invalidLinks;
+
+    if (probableLinksCount === 0) {
+      log(`No probable links found, stopping retry loop after attempt ${attempt}`);
+      break;
+    }
+
+    if (attempt === MAX_ATTEMPTS_PATIENT_LINKING) {
+      log(`Reached maximum retry attempts (${MAX_ATTEMPTS_PATIENT_LINKING}), stopping retry loop`);
+    }
```

Comment:
nits:
- move it outside of the while, so the logic inside is simpler
- defensive programming `if (attempts >= max)`